### ISSUE TITLE
Wavetable Interpolation

### ISSUE THREAD

**Issue by merlina876** [surge-synthesizer/surge#2467]
Would it be possible at some stage to implement an option to interpolate between single cycles in a wavetable and the ability adjust the degree of that interpolation ?
Further more it would be usefull if such a interpolation knob could be modulated by a mseg or lfo or envelope or even assign it to keytrack, velocity, etc.  
Now if you have a full set of 256 different single cycles it changes aprupt and that sounds mostly terrible.

**Comment by baconpaul** [comment]
Doesn’t the morph control on the wavetable oscillator do exactky this? I’m not sure I understand your question

**Comment by merlina876** [comment]
In the original Surge this "Morph" slider had an other name and does not morph but scanning along the wavetable and this is not a morphing / interpolating between the single cycles of a wavetable. 
What I want to request is what Hive 1.2/2 can do. It has a spectral- and amp-interpolation between existing single cycles (beside scanning thru tables), but you can not adjust the degree of it. And this interpolation has nothing to do with that "Morph" - slider in Surge does. 
I hope you see now what I am writting from !

**Comment by baconpaul** [comment]
Not really so let me try again!

If you have a wavetable with 10 frames and you set morph to 3.2, you get an interpolation which is 80% table 3 and 20% table 4. If you modulate it it scans interpolating amongst the frames

How is what you want different than packing your single cycles into a wavetable and using that? That’s kinda what I am not understanding (I don’t have hive) so perhaps dumb it down a bit for me 👍 

Thanks! And interested in understanding. Appreciate your patience

**Comment by baconpaul** [comment]
Ahh one of my colleagues in our slack explained that hive has multiple interpolation methods; snap, smooth, spectral, and zero phase whereas surge only has smooth. Got it! So what I think you want in surge language is for the wavetable morph slider to do something other than linear interpolation? Is that correct?

**Comment by merlina876** [comment]
Let's take the example with 10 single cycles that are in a wavetable.
If you now move that "Morph" -slider in Surge from left 0% to right 100% it scans the existing 10 single cycles from 1-10. That's all. 
If you would now apply an interpolation between every single cycle by 10 additional calculated intrpolation single cycles then this wavetable would have 100 single cycles, 10 which are already in the wavetable and 90 which are calculated in realtime as an interpolation between all this 10 original single cycles. The result would be a much smoother transition between that 10 original single cycles in that wavetable. 
And if you then could modulate such a knob or slider you could get different smootheness in realtime. That would be fantastic. 

I am shure all users would be happy with such a feature !
And of course MSEGs (but you wrote you are already working on it)

**Comment by baconpaul** [comment]
OK here's an example with 16 single cycles in a wavetable from the factory library.

The formula for wavetable n is

```
def sinefm2(x, n):
    c = np.sin(2 * math.pi * x * 2)
    r = np.sin(2 * math.pi * x + (n - 1) * c / 2)
    return r
```

and here's an animated gif of me dragging the morph slider and it interpolating between the frames in a perfectly smooth transition.

What would you like to be different?

![surgetoday](https://user-images.githubusercontent.com/13387561/89842419-8f330a00-db43-11ea-9733-869e61d2831c.gif)

**Comment by merlina876** [comment]
Sorry, I don't see here any animation, but you now know what I think.

There exists different methods for that interpolation, as you wrote above.
For example amplitude-, spectral-, formant-, phase-interpolation etc. 

But as a starting point a good and flexible amplitude-interpolation would be great.
If this would work, you could add some more interpolation methods over the time.

**Comment by baconpaul** [comment]
Right but the code does amplitude interpolation! That’s why I don’t understand. 

Like: here’s the line of code where we interpolate between the two tables

https://github.com/surge-synthesizer/surge/blob/e5105b88e65d48085ef7acdb7d2f8493c7387942/src/common/dsp/WavetableOscillator.cpp#L279

One of us is missing communicating with the other! Because (aside from phase and spectral which are hard) I’m not sure why you can’t do exactly what you want in surge right now!

**Comment by merlina876** [comment]
I wrote this request because I have tested a vocal wavetable with 256 single cycles and the result with this "Morph" -slider in Surge is terrible compared to the morph result in Hive 1.2/2. 
I am not a c++ programmer but the interpolation steps are not enough to get a good result with a wavetable with 256 single cycles.

Would it be possible to increase the number of interpolation steps with big wavetables,
so that the result could be more smooth in that amplitude interpolation. 

This would of course also help many other users and improve the quality of Surge !

**Comment by baconpaul** [comment]
So the interpolation is continuous and bad resolution of a float so there’s a thousand steps beteeen each frame available with 256 frames. But the slider won’t let you get them you in normal mode since it only has 140 pixels! 

But we added typeins and high resolution displays to deal with this.

Can you share the wavetable or patch that is sounding bad? Like I said the interpolation isn’t creating new tables it is doing continuous real-time linear interpolation but with 255 tables you will have to set modulators carefully to make sure to sweep through tables reasonably

**Comment by baconpaul** [comment]
Also with 256 tables if you modulate with (say) a triangle wave if you have a frequency faster than 512/sample rate you will skip tables, plus in surge the lfos linearly interpolate across a block. So if you modulate morph with a 256 frame table you will have to be careful to either have low rate, small modulation depth, or you may skip frames. That math is clear enough I hope. 

But looking at your patch and wavetable could help maybe - if you can attach them here as a zjp

**Comment by mkruselj** [comment]
256 frames wavetables morph perfectly smoothly over here. Don't use the mouse, use host automation.

**Comment by merlina876** [comment]
@baconpaul 
When I have tested  a 256 frame vocal wavetable the vocal table was played back via LFO-Envelope and it stuttert. Maybe the Morph-Slider was in Step-Mode. I have to test it again and will report the result.


@mkruselj 
I have bought Surge years ago and worked some years with the "old" original Surge 1.5.2, therefore I know how to use Surge and all other VSTi I have. Don't regard a person as an Idiot if this is not the case.

I know that different in sound between Surge and Hive, but this is not a question of the generall sound but the playback quality of wavetables in interpolation mode (here called "Morph")
Nevertheless, thanks for the tip.

**Comment by baconpaul** [comment]
The morph slider doesn’t have a step mode; but the LFO rate could skip tables if it is too fast. For instance if you use an envelope which has an attack of 0.05 seconds and you bind it to morph going from 0 - 1 at 48kz you have 44100*0.05/32 values = 68 (since the LFO runs at control frequency of block size) and that could easily lead to skipping of frames.  The wavetable oscillator will linearly interpolate across the change in LFO during a block but still you get the idea - there’s no guarantee that fast modulation in big wavetables will land you in adjacent frames. (The obvious extension of this is imagine modulating morph with a square wave).

But really: If we could see your patch it would help. Like we said we have no problem doing arbitrary interpolation between frames so understanding your specific use case would help.

And I don’t think @mkruselj was regarding you as an idiot; just remind you that surge and hive are different. We’re trying to help you solve your problem here (by first trying to make sure we understand it)

**Comment by mkruselj** [comment]
Just send a patch that you think has a problem (along with an audio example would be nice), you were already asked for this. Thanks.

**Comment by baconpaul** [comment]
Also step mode was UI only; the modulators have always been continuous even in the 1.6s which had the clumsy step gesture. (We upgraded the gesture in 1.7 but really the point is “the oscillator doesn’t care about step mode on the UI control”)

**Comment by mkruselj** [comment]
Also, to clarify on Paul's earlier statement about Morph slider resolution.

The slider internally has the range of 0.0 to 1.0, in 32-bit floating point. That gives us 6 or 7 decimals of precision, which means there is **one million** morphing steps between first and last frame of any wavetable. Which means if your wavetable has 256 frames, we roughly have 3900 steps to go from frame 1 to frame 2.

There is no need to increase this at all, it is perfectly sufficient for all intents and purposes. But yes, send us the patch.

**Comment by merlina876** [comment]
OK, thanks up to now !

I have tested the wavetable playback and interpolation with Surge 1.6.6 and this version had a step entry in the Morph-Mode.
I will now test it again with the latest Surge 1.71 and will attach here the preset file(one or two). The best would be if could use a MSEG for playing this wavetable back. For now I will do a modulation with a LFO and a LFO-Envelope at a slow speed. 
It should be up in some hours

**Comment by baconpaul** [comment]
Yeah the step entry was a ui gesture but didn’t change modulation (which still modulated in beteeen steps). In 1.7 every slider has an option to be quantized if you hold control while dragging, and chooses the appropriate wuantization scale

But again that 16 control didn’t change the sound oscillator or modulation, just the ui Input gesture 

Mseg will be in 1.7.2 for sure it looks like

But we Aren’t sure when 1.7.2 will be ;)

**Comment by merlina876** [comment]
Well, I have done now a second test.

As you can hear in the 2 Serum wavetables the voices sound relative good. This is the quality I hear in Serum. If I load it in Hive 1.2/2 it sounds even better.
I have loaded this two wavetables in Surge with a LFO modulation scanning, 
and you can hear in the Unison 1 presets a slight stutter and cracles. 
In the Unison 4 presets the result is much better especially if you play a chord. 
Therefore I think there could something be improved.
I don't know if the reason lies in the interplation method or something else. Maybe a little phase-interpolation would lead to a better result.
Therefore I would request an additional phase-morph feature !

But what I am now missing in Surge is a "sawtooth up" LFO waveform and of course a mseg would be great for much more flexible mod-curves.


[Surge Voc Demo presets.zip](https://github.com/surge-synthesizer/surge/files/5059986/Surge.Voc.Demo.presets.zip)

**Comment by mkruselj** [comment]
Sawtooth up you get simply by using negative modulation amount...

**Comment by VincyZed** [comment]
I imported the one of the wavetables both in Surge and Serum, automated the position/morph exactly the same and cannot hear a difference between the two..

**Comment by merlina876** [comment]
@mkruselj 
The resynthesis method and results in Serum don't give better results and Serum does not offer a feature to smooth or morph between single cycles. Therefore the apprupt changes in loudnes in the original .wav file can not be smoothed out with a phase-morph. 
But Harmor plays back the original .wav perfect. 
Hive 1.2/2 plays it back best as wavetable with spectral morph mode.


But generally in Surge the quality is still the most worse of all wavetable synths I have and also the quality of filters and effects is rather worse compared to  other VSTi or wavetablesynths I have. Therefore I don't use this VSTi anymore since many years.
But I thought that there has maybe something be improved. The only imprevement I see and hear is the ability to load 2048 single cycles and wavetables with it. 
So, my journey to Surge is now over because it's a waste of time when I have many better VSTis.

Adios

**Comment by K0rrid0r** [comment]
@merlina876 

this is a good moment to actually read the changelog.

https://surge-synthesizer.github.io/changelog/
